### PR TITLE
[DEV-3689] Adds ability to delete Prayer Request

### PR DESCRIPTION
**Pull request by cdanwards**
## DESCRIPTION

As a User, I would like to be able to delete Prayer Requests that I have submitted.

### What does this PR do, or why is it needed?

This PR allows a user to delete a prayer request through the UI. It implements `react-native-action-sheet` which is a new package to take note of. This also adds some tests to the `UserPrayerList` component.

### How do I test this PR?

1. Make sure you have some prayers in `alpha-rock`.
2. yarn
3. yarn start
4. Navigate to the Prayer Menu, then to the `My Prayers` tab.
5. Then select the `...` in the corner of a prayer card and delete!

You can run this query to test to make sure what is in the UI is what's on the server.
```
query {
  getCurrentPersonPrayerRequests {
    id
    firstName
    lastName
    text
    enteredDateTime
  }
}

```

---

> I am affirming this is my _best_ work ([Ecclesiastes 9:10](https://www.bible.com/bible/97/ECC.9.10.MSG))

## TODO

- [x] PR has a relevant title that will be understandable in a public changelog (ie...non developers)
- [x] Closes [DEV-3689](https://newspring.atlassian.net/secure/RapidBoard.jspa?rapidView=71&modal=detail&selectedIssue=DEV-3689)
- [x] No new warnings in tests, in storybook, and in-app
- [ ] Upload GIF(s) of iOS and Android if applicable
- [ ] Set two relevant reviewers

## REVIEW

**Manual QA**

- [ ] Manual QA on iOS and ensure it looks/behaves as expected
- [ ] Manual QA on Android and ensure it looks/behaves as expected

**Code Review: Questions to consider**

- [ ] Read through the "Files changed" tab _very carefully_
- [ ] Edge cases: what assumptions are made about input?
- [ ] What kind of tests could be written?
- [ ] How might we make this easier for someone else to understand?
- [ ] Could the code be simpler?
- [ ] Will the code be easy to modify in the future?
- [ ] What's one part of these changes that makes you excited to merge it?

---

> The purpose of PR Review is to _improve the quality of the software._

### CODE REVIEW

**Comment by redreceipt** [comment]
After deleting, I got a warning saying: "Can't perform a React state update on an unmounted component....memory leak...to fix cancel all subscriptions and asynchronous tasks in `componentWillUnmount`"

Screen Shot 2019-04-23 at 3 07 11 PM

**Comment by redreceipt** [comment]
And when I clicked "Cancel" it deleted the task anyway.

**Comment by redreceipt** [comment]
> After deleting, I got a warning saying: "Can't perform a React state update on an unmounted component....memory leak...to fix cancel all subscriptions and asynchronous tasks in `componentWillUnmount`"
> 

Can't reproduce, this must be unrelated.

**Review by redreceipt — DISMISSED** [review]
Clicking cancel deletes the task too.

**Comment by redreceipt** [comment]
also, my prayers are showing up but I'm not logged in

Untitled Diagram

**Comment by cdanwards** [comment]
@redreceipt FWIW I think the issue of prayers appearing in the prayer list is a byproduct of an error in the `getCurrentPersonPrayerRequests` query, which is outside the scope of this PR. I think we should create a new task for that and go ahead and bring in this PR.

File: packages/newspringchurchapp/src/ui/UserPrayer/UserPrayerCard.js [hunk 1]
```diff
@@ -1,16 +1,23 @@
-import React, { memo } from 'react';
+import React, { PureComponent } from 'react';
 import PropTypes from 'prop-types';
 import moment from 'moment';
+// import { Mutation } from 'react-apollo';
```
Comment (IsaacHardy) [inline packages/newspringchurchapp/src/ui/UserPrayer/UserPrayerCard.js]:
```suggestion

```

File: packages/newspringchurchapp/src/ui/UserPrayer/UserPrayerCard.js [hunk 2]
```diff
@@ -1,16 +1,23 @@
...
   styled,
   CardContent,
   H5,
   BodyText,
   PaddedView,
+  SideBySideView,
+  ButtonLink,
 } from '@apollosproject/ui-kit';
 
+// import deleteUserPrayer from './deleteUserPrayer';
```
Comment (IsaacHardy) [inline packages/newspringchurchapp/src/ui/UserPrayer/UserPrayerCard.js]:
```suggestion

```

File: packages/newspringchurchapp/src/ui/UserPrayer/UserPrayerCard.js [hunk 3]
```diff
@@ -1,16 +1,23 @@
...
+// import getUserPrayers from './getUserPrayers';
```
Comment (IsaacHardy) [inline packages/newspringchurchapp/src/ui/UserPrayer/UserPrayerCard.js]:
```suggestion

```

File: packages/newspringchurchapp/src/ui/UserPrayer/index.js [hunk 4]
```diff
@@ -13,20 +15,48 @@ const StyledView = styled(({ theme }) => ({
...
-          ))
-          .reverse()}
-      </StyledView>
+      <Mutation
+        mutation={deleteUserPrayer}
+        update={async (cache, { data: { deletePublicPrayerRequest } }) => {
+          const currentCurrentPersonPrayerRequests = cache.readQuery({
+            query: getUserPrayers,
+          });
+          const deletedId = deletePublicPrayerRequest.id;
```
Comment (IsaacHardy) [inline packages/newspringchurchapp/src/ui/UserPrayer/index.js]:
```suggestion
          const { id } = deletePublicPrayerRequest;
```

File: packages/newspringchurchapp/src/ui/UserPrayer/index.js [hunk 5]
```diff
@@ -13,20 +15,48 @@ const StyledView = styled(({ theme }) => ({
...
+          const newPrayersList = currentCurrentPersonPrayerRequests.getCurrentPersonPrayerRequests.filter(
+            (prayer) => prayer.id !== deletedId
```
Comment (IsaacHardy) [inline packages/newspringchurchapp/src/ui/UserPrayer/index.js]:
```suggestion
            (prayer) => prayer.id !== id
```

File: packages/newspringchurchapp/src/ui/UserPrayer/index.js [hunk 6]
```diff
@@ -13,20 +15,48 @@ const StyledView = styled(({ theme }) => ({
 }))(View);
 
 const UserPrayerList = () => (
-  <Query query={getUserPrayers} fetchPolicy="cache-and-network">
+  <Query query={getUserPrayers} fetchPolicy="network-only">
```
Comment (IsaacHardy) [inline packages/newspringchurchapp/src/ui/UserPrayer/index.js]: Is this the only way for this query to work as intended? I don't like having to make a call everytime this view is presented.

File: packages/apollos-data-prayer/src/prayer-requests/data-source.js [hunk 7]
```diff
@@ -70,6 +70,19 @@ export default class PrayerRequest extends RockApolloDataSource {
     }
   };
 
+  // MUTATION Delete a prayer request
+  deletePrayer = async (parsedId) => {
+    try {
+      this.expanded = false;
+      const deletedPrayer = await this.getFromId(parsedId);
+      await this.delete(`PrayerRequests/${parsedId}`);
+      this.expanded = true;
```
Comment (IsaacHardy) [inline packages/apollos-data-prayer/src/prayer-requests/data-source.js]: Does this not reset itself when the class is initialized?

File: packages/newspringchurchapp/src/ui/UserPrayer/UserPrayerCard.js [hunk 8]
```diff
@@ -19,22 +26,59 @@ const HeaderView = styled(() => ({
...
-    <CardContent>
-      <PrayerText>{text}</PrayerText>
-    </CardContent>
-  </Card>
-));
+const HorizontalTextLayout = styled(({ theme }) => ({
+  height: theme.helpers.verticalRhythm(0.875),
+}))(SideBySideView);
+
+const cancelIndex = 1;
```
Comment (IsaacHardy) [inline packages/newspringchurchapp/src/ui/UserPrayer/UserPrayerCard.js]: I don't like having this globally. This const should go around line 49.

File: packages/newspringchurchapp/src/ui/UserPrayer/UserPrayerCard.js [hunk 9]
```diff
@@ -19,22 +26,59 @@ const HeaderView = styled(() => ({
...
+const destructiveIndex = 0;
```
Comment (IsaacHardy) [inline packages/newspringchurchapp/src/ui/UserPrayer/UserPrayerCard.js]: I don't like having this globally. This const should go around line 49.

**Review by IsaacHardy — CHANGES_REQUESTED** [review]

File: packages/newspringchurchapp/src/ui/UserPrayer/UserPrayerCard.js [hunk 10]
```diff
@@ -19,22 +22,58 @@ const HeaderView = styled(() => ({
...
+class UserPrayerCard extends PureComponent {
+  handleShowActionSheet = () => {
+    this.ActionSheet.show();
+  };
+
+  render() {
+    const cancelIndex = 1;
+    const destructiveIndex = 0;
+    const { duration, text, id, deletePrayer, ...otherProps } = this.props;
+    const handleOnPress = (index) => {
```
Comment (IsaacHardy) [inline packages/newspringchurchapp/src/ui/UserPrayer/UserPrayerCard.js:38]: @cdanwards Shouldn't this be handled by the ActionSheet? I thought this functionality was built in?

File: packages/newspringchurchapp/src/ui/UserPrayer/UserPrayerCard.js [hunk 11]
```diff
@@ -19,22 +22,58 @@ const HeaderView = styled(() => ({
...
+    const options = ['Remove Prayer Request', 'Cancel'];
+    return (
+      <Card {...otherProps}>
+        <HeaderView>
+          <HorizontalTextLayout>
+            <H5>{moment(duration).fromNow()}</H5>
+            <ButtonLink onPress={this.handleShowActionSheet}>...</ButtonLink>
+            <ActionSheet
+              ref={(o) => {
+                this.ActionSheet = o;
```
Comment (IsaacHardy) [inline packages/newspringchurchapp/src/ui/UserPrayer/UserPrayerCard.js:53]: I have no idea what this does.

File: packages/newspringchurchapp/src/ui/UserPrayer/UserPrayerCard.js (hunk 10 above)
Comment (cdanwards) [reply packages/newspringchurchapp/src/ui/UserPrayer/UserPrayerCard.js:38]: So, I looked into this and apparently the `onPress` event is called no matter which button is pressed, and that's the intended behavior of the package [the creator states that here.](https://github.com/beefe/react-native-actionsheet/issues/12) I just made it so that the `onPress` isn't called for the cancel button.

**Review by IsaacHardy — APPROVED** [review]
